### ISSUE TITLE
lme4: issues with the Gamma family?  'Error in qr.default(X) : NA/NaN/Inf in foreign function call (arg 1)'

### ISSUE THREAD

**Issue by DominiqueMakowski** [rvlenth/emmeans#217]
Hi! Just wanted to report what seems to be a problem running `ref_grid` for mixed models of the Gamma family, that produces this bizarre error as there was a missing name in the data or something. Is there a workaround for gamma models? Thanks a lot!

``` r
library(lme4)
library(emmeans)

data <- iris
data$Petal.Length_factor <- as.factor(ifelse(data$Petal.Length < 4.2, "A", "B"))

# This works
model <- lme4::lmer(Sepal.Width ~ Species + (1 | Petal.Length_factor), data=data)
emmeans::ref_grid(model, at = list("Species" = c("setosa", "versicolor", "virginica")))
#> 'emmGrid' object with variables:
#>     Species = setosa, versicolor, virginica

# This doesn't 
model <- lme4::glmer(Sepal.Width ~ Species + (1 | Petal.Length_factor), data=data, family = "Gamma")
emmeans::ref_grid(model, at = list("Species" = c("setosa", "versicolor", "virginica")))
#> Warning in sqrt(weights): NaNs produced
#> Error in qr.default(X): NA/NaN/Inf in foreign function call (arg 1)
```

<sup>Created on 2020-08-01 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by rvlenth** [comment]
Thanks. I believe this has to do with capabilities just added. I'll check, but I believe this bug won't happen with the latest CRAN version.

Sent from my iPad

On Aug 1, 2020, at 2:44 AM, Dominique Makowski <notifications@github.com> wrote:

﻿

Hi! Just wanted to report what seems to be a problem running ref_grid for mixed models of the Gamma family, that produces this bizarre error as there was a missing name in the data or something. Is there a workaround for gamma models? Thanks a lot!

library(lme4)
library(emmeans)

data <- iris
data$Petal.Length_factor <- as.factor(ifelse(data$Petal.Length < 4.2, "A", "B"))

# This works
model <- lme4::lmer(Sepal.Width ~ Species + (1 | Petal.Length_factor), data=data)
emmeans::ref_grid(model, at = list("Species" = c("setosa", "versicolor", "virginica")))
#> 'emmGrid' object with variables:
#>     Species = setosa, versicolor, virginica

# This doesn't
model <- lme4::glmer(Sepal.Width ~ Species + (1 | Petal.Length_factor), data=data, family = "Gamma")
emmeans::ref_grid(model, at = list("Species" = c("setosa", "versicolor", "virginica")))
#> Warning in sqrt(weights): NaNs produced
#> Error in qr.default(X): NA/NaN/Inf in foreign function call (arg 1)

Created on 2020-08-01 by the reprex package<https://reprex.tidyverse.org> (v0.3.0)

**Comment by rvlenth** [comment]
As I suspected, the problem crops up in a new routine that saves the weighted model matrix (actually just the R part of its QR decomposition) for certain models, for later devious purposes. In this particular example, I retrieved the weights from the `@pp` slot:
```r
> model@pp$Xwts
  [1] -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965
    ...
[141] -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812
```
Obviously bad things happen when I want to compute their square roots.

I have added some code that makes this routine bullet-proof, I hope. With the current example, we now get
```r
> ref_grid(model, at = list("Species" = c("setosa", "versicolor", "virginica")))
'emmGrid' object with variables:
    Species = setosa, versicolor, virginica
Transformation: “inverse” 
```
All looks OK, but behind the scenes, we silently bypassed the problem with the negative weights. The saved model matrix just isn't there but that's OK; we just can't use the new `submodel` option.

Meanwhile, I will try to better understand what is *really* saved in `object@pp$Xwts`.

Thanks again for the report. This I think is my first report of a bug in something that hadn't even yet made it to CRAN!

**Comment by bbolker** [comment]
As commented in e-mail, I believe that these values are actually already the square roots of the weights (but they happen to be the negative square root!); try comparing them to `sqrt(g$weights)` where `g` is the closest analogue `glm()` fit ...

**Comment by rvlenth** [comment]
Thanks, Ben. I will do that comparison. I use this stuff to figure out how to get EMMs for a submodel -- e.g., removing an interaction from the model. The calculations are exact only for an ordinary [weighted] linear regression model, so for a GLM they are already approximate. But still a potentially useful feature. The best way, of course, is to actually fit the submodel.

**Comment by rvlenth** [comment]
As I suspected, the problem crops up in a new routine that saves the weighted model matrix (actually just the R part of its QR decomposition) for certain models, for later devious purposes. In this particular example, I retrieved the weights from the `@pp` slot:
```r
> model@pp$Xwts
  [1] -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965 -3.455965
    ...
[141] -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812 -2.952812
```
Obviously bad things happen when I want to compute their square roots.

I have added some code that makes this routine bullet-proof, I hope. With the current example, we now get
```r
> ref_grid(model, at = list("Species" = c("setosa", "versicolor", "virginica")))
'emmGrid' object with variables:
    Species = setosa, versicolor, virginica
Transformation: “inverse” 
```
All looks OK, but behind the scenes, we silently bypassed the problem with the negative weights. The saved model matrix just isn't there but that's OK; we just can't use the new `submodel` option.

Meanwhile, I will try to better understand what is *really* saved in `object@pp$Xwts`.

Thanks again for the report. This I think is my first report of a bug in something that hadn't even yet made it to CRAN!
